modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/OnflightMessageTracker.java

```diff
@@ -365,6 +383,11 @@ public boolean handleAckReceived(UUID channel, long messageID) throws AndesExcep
             log.debug("Ack Received message id= " + messageID + " channel id= " + channel);
         }
 
+        if (retainedMessageList.contains(messageID)) {
+            retainedMessageList.remove(messageID);
```

Comment:
It will keep messages in retainedMessageList only until ack is received. after that it will remove that message id from the list. It similar to messageSendingTracker concurrent hashmap in OnflightMessageTracker.